### PR TITLE
[FEATURE]: DataProfilerColumnDomainBuilder

### PR DESCRIPTION
**Pull request by alexjeilert**
Changes proposed in this pull request:
- The Capital One Data Profiler Column Domain Builder
- This includes, a class that builds a domain of columns that match a given rule
- Compatible with numeric_rule, text_rule, timestamp_rule, and categorical_rule.

### Definition of Done
Please delete options that are not relevant.

- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/docs/contributing/style_guides/code_style)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/docs/contributing/contributing_checklist) of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added [unit tests](https://docs.greatexpectations.io/docs/contributing/contributing_test#writing-unit-and-integration-tests) where applicable and made sure that new and existing tests are passing.
- [x] I have run any local integration tests and made sure that nothing is broken.

### CODE REVIEW

**Review by anthonyburdi — CHANGES_REQUESTED** [review]
Hi @alexjeilert! Thank you for this contribution. It looks like there are some linting issues, would you mind running `ruff .` and fixing the resulting issues? It looks like some can be auto-fixed by ruff. Thanks! @alexsherstinsky will have more comments on applicability.

File: contrib/capitalone_dataprofiler_expectations/capitalone_dataprofiler_expectations/rule_based_profiler/domain_builder/data_profiler_column_domain_builder.py [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+            column_names=effective_column_names,
+            domain_type=self.domain_type,
+            # type: ignore[union-attr] # could be None
+            table_column_name_to_inferred_semantic_domain_type_map=None,
+            # self.semantic_type_filter.table_column_name_to_inferred_semantic_domain_type_map,
+        )
+
+        return domains
+
+    def get_effective_column_names(
```
Comment (alexsherstinsky) [inline contrib/capitalone_dataprofiler_expectations/capitalone_dataprofiler_expectations/rule_based_profiler/domain_builder/data_profiler_column_domain_builder.py:101]: @alexjeilert This method is provided in the parent class.  It is desirable to use it, but not overwrite it.  There are examples in the core part of the repository.  Do any of them apply to the present case?  Thank you for your consideration.

**Review by alexsherstinsky — CHANGES_REQUESTED** [review]
@alexjeilert I may not be up to speed, so please help me here.  Has this new `DomainBuilder` been requested by CapitalOne, or is it entirely an innovation-inspired effort?  The reason I asked is that as part of my collaboration with CapitalOne, I learned that the idea behind was to have the CapitalOne DataScience team come up with the domains, and the Profiler estimate features based on those domains.  In other words, all of the DataScience and MachineLearning/Estimation part was done by the Open Source CapitalOne Profiler.  Thank you.

File: contrib/capitalone_dataprofiler_expectations/capitalone_dataprofiler_expectations/rule_based_profiler/domain_builder/data_profiler_column_domain_builder.py [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+        variables: Optional[ParameterContainer] = None,
+        rule_name: Optional[str] = None,
+    ) -> List[str]:
+        """
+        This method filters by a rule's criteria to obtain columns to be included as part of returned "Domain" objects.
+        """
+
+        effective_column_names = list()
+
+        rule_name_to_data_types = {
```
Comment (alexsherstinsky) [inline contrib/capitalone_dataprofiler_expectations/capitalone_dataprofiler_expectations/rule_based_profiler/domain_builder/data_profiler_column_domain_builder.py:114]: @alexjeilert A shortcoming of this approach is that the keys in this table are entirely arbitrary.  Perhaps we could come up with a more robust alternative?  Towards this goal, could you please provide a paragraph or two description of the goals that this new "column-kind" `DomainBuilder` should aim to achieve?  Thank you very much (please also feel free to contact me on Slack).

**Comment by alexsherstinsky** [comment]
Closing this pull request, because #7498 provides the required functionality.

**Pull request closed without merging**